core/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -792,6 +793,17 @@ private void parseSource(DefaultSearchContext context, SearchSourceBuilder sourc
             }
             context.storedFieldsContext(source.storedFields());
         }
+
+        if (source.collapse() != null) {
```

Comment:
I think it's ok if we document it ? To me it's the opposite, I like the fact that we don't change the original search request and that the logic is implemented directly in the fetch sub phase but if you feel strongly against it I can replace it with your solution.